File: Carts/Views/CAAuthenticationButton.m
#import <pop/POP.h>
#import <ReactiveCocoa/ReactiveCocoa.h>
#import <MRProgress/MRActivityIndicatorView.h>

#import "CAAuthenticationButton.h"

#import "UIImage+CAImage.h"
#import "CARect.h"
#import "UIColor+CAColor.h"
#import "POPAnimation+CAAnimation.h"
#import "UIView+CAView.h"
#import "CATogCAeAnimator.h"

@interface CAAuthenticationButton ()
@property (nonatomic) BOOL isExtended;

@property (nonatomic) CAAnimation *shake;
@property (nonatomic) CATogCAeAnimator *fadeTogCAeAnimator;

@property (nonatomic) NSString *savedTitle;
@property (nonatomic) UIImage *savedImage;
@property (nonatomic) UIImage *image;
@end

@implementation CAAuthenticationButton

#pragma mark - Init

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        self.isExtended = YES;
        [self setMaskToRoundedCorners:UIRectCornerAllCorners withRadii:4.0];
        self.savedTitle = [self titleForState:UIControlStateNormal];
        self.savedImage = [self imageForState:UIControlStateNormal];
        
        self.image =  [UIImage imageNamed:@"done"];
    }
    
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
  
    [self setMaskToRoundedCorners:UIRectCornerAllCorners withRadii:4.0];
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesEnded:touches withEvent:event];

    [self.fadeTogCAeAnimator togCAeAnimation];
}

- (void)setActivityIndicatorView:(MRActivityIndicatorView *)activityIndicatorView {
    _activityIndicatorView = activityIndicatorView;
    
    self.activityIndicatorView.layer.opacity = 0;
    self.activityIndicatorView.hidesWhenStopped = YES;
    self.activityIndicatorView.hidden = YES;
    
    if (self.shouldAnimate) {
        self.fadeTogCAeAnimator = [CATogCAeAnimator animatorWithTarget:self.activityIndicatorView.layer property:kPOPLayerOpacity startValue:@(0) endValue:@(1)];
        
        @weakify(self);
        self.fadeTogCAeAnimator.forwardsAction = ^{
            @strongify(self);
            [self setTitle:@"" forState:UIControlStateNormal];
            [self setImage:nil forState:UIControlStateNormal];
            [self.activityIndicatorView startAnimating];
        };
        
        self.fadeTogCAeAnimator.backwardsAction = ^{
            @strongify(self);
            [self setTitle:self.savedTitle forState:UIControlStateNormal];
            [self setImage:self.savedImage forState:UIControlStateNormal];
            [self.activityIndicatorView stopAnimating];
        };
    }
}

- (void)animateError {
    POPSpringAnimation *animation = [POPSpringAnimation animationWithPropertyNamed:kPOPLayerPositionX];
    animation.velocity = @2000;
    animation.springBounciness = 20;
    animation.springSpeed = 20;
    
    POPSpringAnimation *ani = [POPSpringAnimation animationWithPropertyNamed:kPOPLayerPositionX];
    ani.velocity = @2000;
    ani.springBounciness = 20;
    ani.springSpeed = 20;
    
    [self.activityIndicatorView.layer pop_addAnimation:animation forKey:@"ani"];
    [self.layer pop_addAnimation:ani forKey:@"aani"];
    
    [self.fadeTogCAeAnimator togCAeAnimation];
}

- (void)animateSuccess {
    [self.activityIndicatorView stopAnimating];
    [self setTitle:@"" forState:UIControlStateNormal];
    [self setImage:self.image forState:UIControlStateNormal];
}

@end

File: Carts/Views/CAAuthenticationButton.swift
//
//  CAAuthenticationButton.swift
//  Carts
//
//  Created by Hakon Hanesand on 4/26/15.
//  Copyright (c) 2015 Hakon Hanesand. All rights reserved.
//

import UIKit

class CAAuthenticationButton: UIButton {
    @IBOutlet weak var activityIndicator: MRActivityIndicatorView! {
        didSet {
            if self.shouldAnimate! {
                self.fade = CAToggleAnimator(target: self.activityIndicator.layer, property: kPOPLayerOpacity, start: 0, end: 1)
                
                self.fade?.forwardsAction = {
                    self.setTitle(nil, forState: .Normal)
                    self.setImage(nil, forState: .Normal)
                    self.activityIndicator.startAnimating()
                }
                
                self.fade?.backwardsAction = {
                    self.setTitle(self.savedTitle, forState: .Normal)
                    self.setImage(self.savedImage, forState: .Normal)
                    self.activityIndicator.startAnimating()
                }
            }
        }
    }
    
    @IBInspectable var shouldAnimate: Bool?
    
    var shake: CAStoredAnimation?
    var fade: CAToggleAnimator?
    
    var savedTitle: String?
    var savedImage: UIImage?
    
    var image: UIImage

    required init(coder aDecoder: NSCoder) {
        self.image = UIImage(named: "done")!
        super.init(coder: aDecoder)
    }
    
    func animateError() {
        let posx = POPSpringAnimation(propertyNamed: kPOPLayerPositionX)
        posx.velocity = 2000;
        posx.springBounciness = 20;
        posx.springSpeed = 20;
        
        let posx2 = POPSpringAnimation(propertyNamed: kPOPLayerPositionX)
        posx2.velocity = 2000;
        posx2.springBounciness = 20;
        posx2.springSpeed = 20;
        
        self.activityIndicator.layer.pop_addAnimation(posx, forKey: "ani")
        self.layer.pop_addAnimation(posx2, forKey: "ani2")
        
        self.fade?.toggleAnimation()
    }
    
    func animateSuccess() {
        self.activityIndicator.startAnimating()
        self.setTitle(nil, forState: .Normal)
        self.setImage(self.image, forState: .Normal)
    }
}
